### PR TITLE
Allow infallible construction of `Signature`

### PR DESCRIPTION
**Pull request by yukibtc**
Hi! This PR add 1 constructor and 2 methods to `Signature` struct:

* Add `Signature::from_byte_array` constructor
* Add `Signature::to_byte_array` and `Signature::as_byte_array` methods
* Deprecate `Signature::serialize` method

### CODE REVIEW

**Comment by Kixunil** [comment]
Also the title is misleading since this is not what "zero copy" means. "Zero copy" is just a stupid name for "zero heap allocations" and the original code didn't do any heap allocations in the first place. Further, in `hashes` we use `from_byte_array` for array-accepting constructors.

What I believe is non-controversial about this PR: we want *some* infallible method that accepts an array. I wanted to suggest you to make just that change first but then I remembered about `from_byte_array` thing.

**Comment by apoelstra** [comment]
Neat! Concept ACK.

I would like to use `to_byte_array` and `from_byte_array` to be consistent with `hashes`. (I have proposed the same thing for `SecretKey` in #724.)

**Comment by yukibtc** [comment]
> Also the title is misleading since this is not what "zero copy" means. "Zero copy" is just a stupid name for "zero heap allocations" and the original code didn't do any heap allocations in the first place

Ahh, thanks for the clarification

---

I've done the changes as suggested.

File: src/schnorr.rs [hunk 1]
```diff
@@ -88,9 +91,17 @@ impl Signature {
     }
 
     /// Returns a signature as a byte array.
-    #[inline]
+    #[deprecated(since = "0.30.0", note = "Use `to_byte_array` instead.")]
```
Comment (Kixunil) [inline src/schnorr.rs:94]: We usually use `TBD` in since and mass-replace it during release but there isn't anything wrong with this.

**Review by Kixunil — APPROVED** [review]
ACK 23b87a61773edd2b2897b92ca547a84f433dab8d

**Review by apoelstra — APPROVED** [review]
ACK 23b87a61773edd2b2897b92ca547a84f433dab8d successfully ran local tests

**Pull request merged**